### PR TITLE
[Snappi] Handling multi-dut scenario for verify_pause_frame_count in PFC lossless test

### PR DESCRIPTION
**Pull request by amitpawar12**
### Description of PR
The PFC multi-dut testcases checks the pause frame count transmitted and received by the DUT during the test.

Testcase needed modifications to handle the multiple line card scenario.

Summary:
Fixes # (issue)
Issue #12887 

### Type of change

- [X] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [ ] Test case(new/improvement)

### Back port request
- [ ] 202012
- [ ] 202205
- [ ] 202305
- [ ] 202311
- [X] 202405

#### What is the motivation for this PR?
The PFC lossless testcases uses multidut_helper file to start the traffic and verify the results. One of the verifications involves checking the pause frames being sent and received by the DUT.

A single DUT (object duthost) was being sent to the function 'verify_pause_frame_count' to verify the sent and received pause frames. This will work fine for - Single line card single ASIC and Single line card multiple ASIC - because it is single DUT only.

However, this would fail for multiple line card scenario. Here, the PFC receiving and sending line cards are different, and hence the failure.

#### How did you do it?
In the multidut_helper file, there is now a code to identify 'rx_dut' (Ingress DUT receiving packets from IXIA) and 'tx_dut' (Egress DUT sending packets to IXIA). 

Instead of sending single DUT variable to verify_pause_frame_count, it is now passing rx_dut and tx_dut to it. The 'verify_pause_frame_count' will check the incoming pause frame counts on tx_dut and outgoing pause frame counts  on rx_dut.

Since line card is same in case of single line card, the 'rx_dut' and 'tx_dut' will be same and hence will continue to work as is.

#### How did you verify/test it?
Verified with multiple line card scenario test with both 100 and 400Gbps ingress and egress line cards.

Code snapshot:
```
20:00:16 traffic_generation.verify_pause_frame_co L1509 INFO   | Rx Pause Frames:2154763 on DUT:ixre-egl-board72, port:Ethernet48 and Prio:3
20:00:18 traffic_generation.verify_pause_frame_co L1509 INFO   | Rx Pause Frames:2154763 on DUT:ixre-egl-board72, port:Ethernet48 and Prio:4
20:00:19 traffic_generation.verify_pause_frame_co L1528 INFO   | Tx Pause Frames:1371912 on DUT:ixre-egl-board71, port:Ethernet40 and Prio:3
20:00:20 traffic_generation.verify_pause_frame_co L1528 INFO   | Tx Pause Frames:1371912 on DUT:ixre-egl-board71, port:Ethernet40 and Prio:4
```

### CODE REVIEW

**Review by abdosi — APPROVED** [review]

**Comment by abdosi** [comment]
@sdszhang : Please review

File: tests/common/snappi_tests/traffic_generation.py [hunk 1]
```diff
@@ -578,7 +578,8 @@ def verify_in_flight_buffer_pkts(duthost,
                               format(dropped_packets))
 
 
-def verify_pause_frame_count_dut(duthost,
+def verify_pause_frame_count_dut(rx_dut,
```
Comment (sdszhang) [inline tests/common/snappi_tests/traffic_generation.py:581]: This function is also used in tests/snappi_tests/pfc/files/helper.py. Maybe change that part too.

File: tests/common/snappi_tests/traffic_generation.py [hunk 2]
```diff
@@ -587,7 +588,8 @@ def verify_pause_frame_count_dut(duthost,
     on the DUT
 
     Args:
-        duthost (obj): DUT host object
+        rx_dut (obj): Ingress DUT host object
+        tx_dut (obj): Egress DUT host object
```
Comment (sdszhang) [inline tests/common/snappi_tests/traffic_generation.py:592]: pls reuse the same comment here too to indicate what traffic are expected on rx and tx ports.
# rx_dut is ingress DUT receiving packets.
# tx_dut is egress DUT transmitting packets and also receiving PFCs.

File: tests/snappi_tests/multidut/pfc/files/multidut_helper.py [hunk 3]
```diff
@@ -80,6 +80,14 @@ def run_pfc_test(api,
     duthost2 = snappi_extra_params.multi_dut_params.duthost2
     tx_port = snappi_extra_params.multi_dut_params.multi_dut_ports[1]
 
+    # rx_dut is ingress DUT receiving packets.
+    # tx_dut is egress DUT transmitting packets and also receiving PFCs.
```
Comment (sdszhang) [inline tests/snappi_tests/multidut/pfc/files/multidut_helper.py:84]: very good comments!

**Comment by amitpawar12** [comment]
Thanks for the comment - @sdszhang.

I have made changes to snappi_tests/pfc/files/helper.py as well to account for the changes in verify_pause_frame_count.

Additionally, 'verify_egress_queue_frame_count' function in multidut_helper file also should be checking the egress_queue_frame_count on the egress DUT(tx_dut) and hence changed it accordingly.

Thanks,
-A

**Review by sdszhang — APPROVED** [review]

**Comment by sdszhang** [comment]
@amitpawar12 can you fix the prep-commit failure?

File: tests/common/snappi_tests/traffic_generation.py (hunk 1 above)
Comment (amitpawar12) [reply tests/common/snappi_tests/traffic_generation.py:581]: Yes. I have made changes to helper.py as well.

**Comment by amitpawar12** [comment]
Fixed the pre-commit issue. 

Thanks,
-A

**Review by yejianquan — APPROVED** [review]
LGTM

**Pull request merged**